account-integrations/safe/test/hardhat/integration/SafeZKPPasswordPlugin.test.ts

```diff
@@ -0,0 +1,194 @@
+      throw new Error("No entry points found");
+    }
+
+    const ssf = await SafeSingletonFactory.init(userWallet);
+    const zkpClient = await ERC4337ZKPPasswordClient.create();
+
+    return {
+      factory: await ssf.connectOrDeploy(SafeProxyFactory__factory, []),
+      singleton: await ssf.connectOrDeploy(Safe__factory, []),
+      bundlerProvider,
+      provider,
+      userWallet,
+      entryPoints,
+      zkpClient,
+    };
+  };
+
+  /**
+   * This test verifies a ERC4337 transaction succeeds when sent via a plugin
+   * The user operation deploys a Safe with the ERC4337 plugin and a handler
+   * and executes a transaction, thus verifying two things:
+   * 1. Deployment of the Safe with the ERC4337 plugin and handler is possible
+   * 2. Executing a transaction is possible
+   */
+  itif("should pass the ERC4337 validation", async () => {
+    const {
+      singleton,
+      provider,
+      bundlerProvider,
+      userWallet,
+      entryPoints,
+      zkpClient,
+    } = await setupTests();
+    const ENTRYPOINT_ADDRESS = entryPoints[0];
+
+    const ssf = await SafeSingletonFactory.init(userWallet);
+
+    const safeZKPPasswordFactory = await ssf.connectOrDeploy(
+      SafeZKPPasswordFactory__factory,
+      [],
+    );
+
+    const signer = await provider.getSigner();
+    // TODO (merge-ok) Use real verifier from zkp dir
+    // https://github.com/getwax/wax/issues/143
+    const passwordVerifier = await new MockPasswordVerifier__factory(
+      signer,
+    ).deploy();
+
+    const feeData = await provider.getFeeData();
+    if (!feeData.maxFeePerGas || !feeData.maxPriorityFeePerGas) {
+      throw new Error(
+        "maxFeePerGas or maxPriorityFeePerGas is null or undefined",
+      );
+    }
+
+    const maxFeePerGas = `0x${feeData.maxFeePerGas.toString()}`;
+    const maxPriorityFeePerGas = `0x${feeData.maxPriorityFeePerGas.toString()}`;
+
+    const owner = ethers.Wallet.createRandom();
+
+    const createArgs = [
+      singleton,
+      ENTRYPOINT_ADDRESS,
+      owner.address,
+      0,
+      passwordVerifier,
+    ] satisfies Parameters<typeof safeZKPPasswordFactory.create.staticCall>;
+
+    const accountAddress = await safeZKPPasswordFactory.create.staticCall(
+      ...createArgs,
+    );
+
+    await receiptOf(safeZKPPasswordFactory.create(...createArgs));
+
+    const to = "0x42ef9B061d2B8416387FaA738Af7251668b0b142"; // Random address
+    const value = ethers.parseEther("1");
+    const data = "0x"; // ETH transfer
+
+    const userOpCallData =
+      SafeZKPPasswordPlugin__factory.createInterface().encodeFunctionData(
+        "execTransaction",
+        [to, value, data],
+      );
+
+    // Native tokens for the pre-fund 💸
+    await receiptOf(
+      userWallet.sendTransaction({
+        to: accountAddress,
+        value: ethers.parseEther("100"),
+        // Re-sync nonce
+        // TODO (merge-ok) See if NonceManager could fix this.
+        nonce: await userWallet.getNonce(),
```

Comment:
@JohnGuilding This is the only spot in my tests that had nonce issues. Looks like `userWallet` is desyncing somewhere above.

Reply:
`userWallet` uses the default mnemonic by the look of things. When using the default mnemonic for the admin signer in my changes, I was getting nonce too low errors too, and resolved those with a nonce manager. Imagine wrapping it in a `NonceManager` will resolve this for you too.

It may be clashing with your local bundler if that uses the following mnemonic: `test test test test test test test test test test test junk` (which the default config for eth-infinitism's bundler does). I've just pushed a commit to my PR that avoids using the first derived account from that mnemonic to avoid clashes with  a bundler

Reply:
I ran into this too and put this list together so I didn't need to deal with real-looking seedphrases during development: [Test bip39 seedphrases](https://gist.github.com/voltrevo/df7c306d0ccad3bbff3cec95a3ec752a).